src/rt/aaA.d

```diff
@@ -114,442 +263,407 @@ TypeInfo_Struct tiEntry(const TypeInfo keyti, const TypeInfo valti)
 }
 
-// bug 13748
-pure nothrow unittest
+/// foreach opApply over all key/value pairs
+extern (C) int _aaApply2(AA aa, in size_t keysz, dg2_t dg)
 {
-    int[int] aa;
-    // make all values go into the last bucket (int hash is simply the int)
-    foreach(i; 0..16)
+    if (aa.empty)
+        return 0;
+
+    immutable off = aa.keysz + aa.keypad;
+    foreach (b; aa.buckets)
     {
-        aa[3 + i * 4] = 1;
-        assert(aa.keys.length == i+1);
+        if (!b.filled)
+            continue;
+        if (auto res = dg(b.entry, b.entry + off))
+            return res;
     }
-
-    // now force a rehash, but with a different value
-    aa[0] = 1;
-    assert(aa.keys.length == 17);
+    return 0;
 }
 
-
-/*************************************************
- * Get pointer to value in associative array indexed by key.
- * Returns null if it is not already there.
- */
-inout(void)* _aaGetRvalueX(inout AA aa, in TypeInfo keyti, in size_t valuesize, in void* pkey)
+/// Construct an associative array of type ti from keys and value
+extern (C) Impl* _d_assocarrayliteralTX(const TypeInfo_AssociativeArray ti, void[] keys,
+    void[] vals)
 {
-    return _aaInX(aa, keyti, pkey);
-}
+    assert(keys.length == vals.length);
 
+    immutable keysz = ti.key.tsize;
+    immutable valsz = ti.value.tsize;
+    immutable length = keys.length;
 
-/*************************************************
- * Determine if key is in aa.
- * Returns:
- *      null    not in aa
- *      !=null  in aa, return pointer to value
- */
-inout(void)* _aaInX(inout AA aa, in TypeInfo keyti, in void* pkey)
-in
-{
-}
-out (result)
-{
-    //assert(result == 0 || result == 1);
-}
-body
-{
-    if (aa.impl is null)
+    if (!length)
         return null;
 
-    //printf("_aaIn(), .length = %d, .ptr = %x\n", aa.a.length, cast(uint)aa.a.ptr);
-    if (immutable len = aa.impl.buckets.length)
+    auto aa = new Impl(ti, nextpow2(INIT_DEN * length / INIT_NUM));
+
+    void* pkey = keys.ptr;
+    void* pval = vals.ptr;
+    immutable off = aa.keysz + aa.keypad;
+    foreach (_; 0 .. length)
     {
-        immutable key_hash = keyti.getHash(pkey);
-        immutable i = key_hash % len;
-        //printf("hash = %d\n", key_hash);
+        immutable hash = calcHash(pkey, ti.key);
 
-        inout(Entry)* e = aa.impl.buckets[i];
-        while (e !is null)
+        auto p = aa.findSlotLookup(hash, pkey, ti.key);
+        if (p is null)
         {
-            if (key_hash == e.hash)
-            {
-                if (keyti.equals(pkey, e + 1))
-                    return cast(inout void*)(e + 1) + aligntsize(keyti.tsize);
-            }
-            e = e.next;
+            p = aa.findSlotInsert(hash);
+            p.hash = hash;
+            p.entry = allocEntry(aa, pkey);
+            aa.firstUsed = min(aa.firstUsed, cast(uint)(p - aa.buckets.ptr));
         }
-    }
+        // set hash and blit value
+        auto pdst = p.entry + off;
+        pdst[0 .. valsz] = pval[0 .. valsz];
```

Comment:
Done, that actually revealed a bug in the existing AA.
When values get overwritten during AA literal construction they need to be destroyed.